File: app/src/main/java/com/njair/homeledger/ui/members/MembersFragment.java
package com.njair.homeledger.ui.members;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.InputType;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.njair.homeledger.GroupMain;
import com.njair.homeledger.R;
import com.njair.homeledger.Service.Group;
import com.njair.homeledger.Service.Member;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import petrov.kristiyan.colorpicker.ColorPicker;

public class MembersFragment extends Fragment {

    private String roomKey;

    public MembersAdapter membersAdapter;
    private TextView textViewDisplay;
    private TextView textViewDisplay2;
    private boolean hasDownloaded = false;

    public Resources r;

    public Group group;
    public DatabaseReference groupRef;

    private MembersViewModel membersViewModel;
    private final int[] colorList = {Color.RED, Color.BLUE, Color.rgb(0, 0, 128),
            Color.rgb(204, 0, 204), Color.rgb(0, 128, 0), Color.rgb(204,204,0),
            Color.rgb(255, 165, 0), Color.rgb(128,0,128), Color.rgb(204, 0, 102), Color.rgb(218,165,32)};

    public View onCreateView(@NonNull final LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        membersViewModel =
                ViewModelProviders.of(this).get(MembersViewModel.class);
        View root = inflater.inflate(R.layout.fragment_members, container, false);
        /*final TextView textView = root.findViewById(R.id.text_notifications);
        membersViewModel.getText().observe(this, new Observer<String>() {
            @Override
            public void onChanged(@Nullable String s) {
                textView.setText(s);
            }
        });*/

        r = getResources();

        roomKey = ((GroupMain) getActivity()).getRoomKey();
        groupRef = FirebaseDatabase.getInstance().getReference().child("groups").child(roomKey);

        RecyclerView membersRecyclerView = root.findViewById(R.id.recyclerview_members);
        membersAdapter = new MembersAdapter(getActivity());
        membersRecyclerView.setAdapter(membersAdapter);
        membersRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false));
        membersRecyclerView.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL));

        textViewDisplay = root.findViewById(R.id.textViewDisplay);
        textViewDisplay2 = root.findViewById(R.id.textViewDisplay2);
        textViewDisplay.setVisibility(View.INVISIBLE);
        textViewDisplay2.setVisibility(View.INVISIBLE);

        //region [Manage FAB]
        final FloatingActionButton fabAdd = root.findViewById(R.id.fabAdd);
        fabAdd.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View view) {
                if(group.memberList.size() < colorList.length){
                    AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
                    builder.setTitle(r.getString(R.string.add_member));

                    final EditText input = new EditText(getActivity());
                    input.setHint(r.getString(R.string.member_name));
                    input.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PERSON_NAME);
                    builder.setView(input);

                    builder.setPositiveButton(r.getString(R.string.add), new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            String inputName = input.getText().toString().trim();

                            if(checkNameAvailable(inputName))
                                new Member(inputName, randomAvailableColor()).upload(roomKey, null);
                            else
                                Snackbar.make(view, r.getString(R.string.member_name_not_available), Snackbar.LENGTH_LONG).show();
                        }
                    });
                    builder.setNegativeButton(r.getString(R.string.cancel), new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {}
                    });

                    builder.show();
                    }
                else Snackbar.make(view, getString(R.string.maximum_amount_of_members_reached), Snackbar.LENGTH_LONG).show();
            }
        });

        membersRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (dy > 0 && fabAdd.getVisibility() == View.VISIBLE) {
                    fabAdd.hide();
                } else if (dy < 0 && fabAdd.getVisibility() != View.VISIBLE) {
                    fabAdd.show();
                }
            }
        });
        //endregion

        return root;
    }

    public void setGroup(Group group){
        this.group = group;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    private boolean checkColorAvailable(int color){
        for(Member member : group.memberList){
            if(member.color == color)
                return false;
        }

        return true;
    }

    private boolean checkNameAvailable(String name){
        for(Member member : group.memberList){
            if(member.getName().trim().equals(name.trim())){
                return false;
            }
        }

        return true;
    }

    private int randomAvailableColor(){
        int colorListLength = colorList.length;

        if(group.memberList.size() < colorListLength){
            int color;
            int randomInt = new Random().nextInt(colorListLength);

            for(int i = randomInt; i != randomInt - 1; i = (i + 1) % colorListLength){
                if(checkColorAvailable(colorList[i]))
                    return colorList[i];
            }

            if(checkColorAvailable(colorList[randomInt - 1]))
                return colorList[randomInt - 1];
        }

        return Color.BLACK;
    }

    public class MembersAdapter extends RecyclerView.Adapter<MembersAdapter.ViewHolder> {
        private Context context;

        public MembersAdapter(Context context) {
            this.context = context;
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(context).inflate(R.layout.layout_member, parent, false);
            ViewHolder viewHolder = new ViewHolder(view);
            return viewHolder;
        }

        @SuppressLint("ClickableViewAccessibility") //TODO ClickableViewAccessibility
        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
            //Toast.makeText(getActivity(), "onBindViewHolder: called", Toast.LENGTH_SHORT).slideUp();
            //members.get(position).adaptView((TransactionLayout) holder.view);

            final Member member = group.memberList.get(position);

            holder.textViewMemberName.setText(member.getName());
            drawableChangeTint(holder.textViewMemberName.getCompoundDrawables(), member.color);

            //region [Handle imageViewTrash: delete member]
            final ImageView IVTrash = holder.imageViewTrash;
            IVTrash.setColorFilter(member.color);
            IVTrash.setOnTouchListener(new View.OnTouchListener() {
                @Override
                public boolean onTouch(View v, MotionEvent event) {
                    switch (event.getAction()) {
                        case MotionEvent.ACTION_DOWN: {
                            IVTrash.setColorFilter(member.color/3);
                            IVTrash.invalidate();
                            break;
                        }
                        case MotionEvent.ACTION_UP:
                            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
                            builder.setMessage(String.format(r.getString(R.string.delete_member_), member.getName()))
                                    .setPositiveButton(r.getString(R.string.yes), new DialogInterface.OnClickListener() {
                                        public void onClick(DialogInterface dialog, int id) {
                                            if(member.getNodeId() != null && !member.getNodeId().equals("")){
                                                groupRef.child("members").child(member.getNodeId()).removeValue().addOnCompleteListener(new OnCompleteListener<Void>() {
                                                    @Override
                                                    public void onComplete(@NonNull Task<Void> task) {
                                                        Snackbar.make(requireActivity().findViewById(android.R.id.content),
                                                                String.format(r.getString(R.string.member__deleted), member.getName()), Snackbar.LENGTH_LONG)
                                                                .setAction(r.getString(R.string.undo), new View.OnClickListener() {
                                                                    @Override
                                                                    public void onClick(View v) {
                                                                        member.upload(roomKey, null);
                                                                    }
                                                                })
                                                                .setActionTextColor(member.color/3)
                                                                .show();
                                                    }
                                                });
                                            } else {
                                                Toast.makeText(context, r.getString(R.string.an_error_has_occurred) + "member doesn't exist in database, try refreshing.",
                                                        Toast.LENGTH_LONG).show();
                                            }
                                        }
                                    })
                                    .setNegativeButton(r.getString(R.string.no), new DialogInterface.OnClickListener() {
                                        public void onClick(DialogInterface dialog, int id) {}
                                    }).show();

                        case MotionEvent.ACTION_CANCEL: {
                            IVTrash.setColorFilter(member.color);
                            IVTrash.invalidate();
                            break;
                        }
                    }

                    return false;
                }
            });
            //endregion

            //region [Handle imageViewBrush: change member color]
            final ImageView IVBrush = holder.imageViewBrush;
            IVBrush.setColorFilter(member.color);
            IVBrush.setOnTouchListener(new View.OnTouchListener() {
                @Override
                public boolean onTouch(View v, MotionEvent event) {
                    switch (event.getAction()) {
                        case MotionEvent.ACTION_DOWN:{
                            IVBrush.setColorFilter(member.color/3);
                            IVBrush.invalidate();
                            break;}

                        case MotionEvent.ACTION_UP:
                            IVBrush.setColorFilter(member.color);
                            IVBrush.invalidate();

                            if(group.memberList.size() < colorList.length){
                                ColorPicker colorPicker = new ColorPicker(getActivity());
                                colorPicker.setTitle(String.format(getResources().getString(R.string.choose_a_color_for_), member.getName()));
                                colorPicker.setColors(removeColorsFromArrayToHex(colorList));
                                colorPicker.setOnFastChooseColorListener(new ColorPicker.OnFastChooseColorListener() {
                                    @Override
                                    public void setOnFastChooseColorListener(int i, int color) {
                                        if(color != Color.WHITE){
                                            member.setColor(color);
                                            member.upload(roomKey, null);
                                        }
                                    }

                                    @Override
                                    public void onCancel() {}
                                });

                                colorPicker.setRoundColorButton(true);
                                colorPicker.show();
                            } else
                                Snackbar.make(v, r.getString(R.string.no_colors_are_available), Snackbar.LENGTH_LONG).show();

                            break;

                        case MotionEvent.ACTION_CANCEL:
                            IVBrush.setColorFilter(member.color);
                            IVBrush.invalidate();
                            break;
                    }

                    return false;
                }
            });
            //endregion
        }

        private ArrayList<String> removeColorsFromArrayToHex(int[] list){
            ArrayList<String> newList = new ArrayList<>();

            for(int i = 0; i < list.length; i++){
                if(checkColorAvailable(list[i]))
                    newList.add(String.format("#%06X", (0xFFFFFF & list[i])));
            }

            return newList;
        }

        @Override
        public int getItemCount() {
            int size = group.memberList.size();
            if(size == 0 && hasDownloaded){
                textViewDisplay.setVisibility(View.VISIBLE);
                textViewDisplay2.setVisibility(View.VISIBLE);

            }
            else{
                textViewDisplay.setVisibility(View.GONE);
                textViewDisplay2.setVisibility(View.GONE);
            }

            return group.memberList.size();
        }

        public class ViewHolder extends RecyclerView.ViewHolder {
            TextView textViewMemberName;
            ImageView imageViewTrash;
            ImageView imageViewBrush;

            public ViewHolder(View itemView) {
                super(itemView);
                textViewMemberName = itemView.findViewById(R.id.textViewMemberName);
                imageViewTrash = itemView.findViewById(R.id.imageViewTrash);
                imageViewBrush = itemView.findViewById(R.id.imageViewBrush);
            }
        }

        private void drawableChangeTint(Drawable[] drawables, int color) {
            for (Drawable drawable : drawables) {
                if (drawable != null)
                    drawable.setTint(color);
            }
        }
    }
}

File: app/src/main/java/com/njair/homeledger/ui/fragments/MembersFragment.kt
package com.njair.homeledger.ui.fragments

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.database.DatabaseReference
import com.njair.homeledger.GroupMain
import com.njair.homeledger.R
import com.njair.homeledger.extensions.Database
import com.njair.homeledger.extensions.Util.Companion.colors
import com.njair.homeledger.extensions.bindView
import com.njair.homeledger.extensions.toHex
import com.njair.homeledger.service.Group
import com.njair.homeledger.service.Member
import com.njair.homeledger.service.OnTouchImageViewListener
import petrov.kristiyan.colorpicker.ColorPicker
import petrov.kristiyan.colorpicker.ColorPicker.OnFastChooseColorListener
import java.util.*
import kotlin.collections.ArrayList

class MembersFragment : Fragment() {
    private val activity: Activity by lazy { requireActivity() }
    private lateinit var roomKey: String
    
    val membersAdapter: MembersAdapter by lazy { MembersAdapter(activity) }
    
    private lateinit var root: View
    private val textViewDisplay: TextView  by lazy { root.findViewById<TextView>(R.id.textViewDisplay) }
    private val textViewDisplay2: TextView by lazy { root.findViewById<TextView>(R.id.textViewDisplay2) }
    
    private lateinit var group: Group
    private val colorList = colors
    
    private var groupRef: DatabaseReference? = null
    
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        root = inflater.inflate(R.layout.fragment_members, container, false)

        roomKey = (activity as GroupMain).roomKey
        groupRef = Database.REFERENCE.child(Database.GROUPS).child(roomKey)
        
        val membersRecyclerView: RecyclerView = root.findViewById(R.id.recyclerview_members)
        
        membersRecyclerView.adapter = membersAdapter
        membersRecyclerView.layoutManager = LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false)
        membersRecyclerView.addItemDecoration(DividerItemDecoration(requireContext(), DividerItemDecoration.VERTICAL))

        //region [Manage FAB]
        val fabAdd: FloatingActionButton = root.findViewById(R.id.fabAdd)
        
        fabAdd.setOnClickListener { view ->
            if (group.memberList.size < colorList.size)
                AlertDialog.Builder(activity).apply {
                    setTitle(resources.getString(R.string.add_member))
    
                    val input = EditText(activity)
                    input.hint = resources.getString(R.string.member_name)
                    input.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PERSON_NAME
    
                    setView(input)
                    setPositiveButton(resources.getString(R.string.add)) { _, _ ->
                        val inputName = input.text.toString().trim { it <= ' ' }
        
                        if (checkNameAvailable(inputName))
                            Member(inputName, randomAvailableColor()).upload(roomKey)
                        else
                            Snackbar.make(view, resources.getString(R.string.member_name_not_available), Snackbar.LENGTH_LONG).show()
                    }
                    setNegativeButton(resources.getString(R.string.cancel), null)
                    show()
                }
            else
                Snackbar.make(view, getString(R.string.maximum_amount_of_members_reached), Snackbar.LENGTH_LONG).show()
        }
        
        membersRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                
                if (dy > 0 && fabAdd.visibility == View.VISIBLE)
                    fabAdd.hide()
                else if (dy < 0 && fabAdd.visibility != View.VISIBLE)
                    fabAdd.show()
            }
        })
        //endregion
        return root
    }

    fun setGroup(group: Group) {
        if(this::group.isInitialized)
            this.group.detach()
        
        this.group = group
        roomKey = group.key
    }

    fun updateUI() {
        val displayVisibility = if (membersAdapter.itemCount == 0) View.VISIBLE else View.GONE
        
        textViewDisplay.visibility  = displayVisibility
        textViewDisplay2.visibility = displayVisibility
    }

    private fun checkColorAvailable(color: Int): Boolean {
        for (member: Member in group.memberList) {
            if (member.color == color)
                return false
        }
        return true
    }

    private fun checkNameAvailable(name: String): Boolean {
        for (member: Member in group.memberList) {
            if ((member.name.trim { it <= ' ' } == name.trim { it <= ' ' })) {
                return false
            }
        }
        return true
    }

    private fun randomAvailableColor(): Int {
        val colorListLength = colorList.size
        if (group.memberList.size < colorListLength) {
            val randomInt = Random().nextInt(colorListLength)
            var i = randomInt
            while (i != randomInt - 1) {
                if (checkColorAvailable(colorList[i]))
                    return colorList[i]
                
                i = (i + 1) % colorListLength
            }
            
            if (checkColorAvailable(colorList[randomInt - 1]))
                return colorList[randomInt - 1]
        }
        return Color.BLACK
    }

    inner class MembersAdapter(private val context: Context) : RecyclerView.Adapter<MembersAdapter.ViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view = LayoutInflater.from(context).inflate(R.layout.layout_member, parent, false)
            return ViewHolder(view)
        }

        @SuppressLint("ClickableViewAccessibility") //TODO ClickableViewAccessibility
        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val member = group.memberList[position]
            val iVLink = holder.imageViewLink
            member.invalidateDrawable()
            //holder.imageViewMemberIcon.setImageResource(getMemberIcon(member.type))
            holder.imageViewMemberIcon.setImageDrawable(member.drawable)
//            holder.imageViewMemberIcon.setColorFilter(member.color)
            holder.imageViewMemberIcon.postInvalidate()
            
            holder.imageViewMemberStar.visibility = if ((member.isAdmin)) View.VISIBLE else View.GONE
            holder.imageViewMemberStar.setColorFilter(member.color * 2 / 3)
            
            holder.textViewMemberName.text = member.name

            //region [Handle imageViewLink: link member to user]
            if (member.user.uid != null && member.user.uid != "")
                iVLink.visibility = View.GONE
            
            iVLink.setColorFilter(member.color)
            iVLink.setOnTouchListener(object : OnTouchImageViewListener(member.color) {
                override fun onTrigger() {
                    AlertDialog.Builder(activity)
                        .setMessage(String.format("Do you want to link the member '%s' with your user account?", member.name)) //TODO Extract string
                        .setPositiveButton(R.string.ok) { _, _ ->
                            //TODO Handle Link Event
                        }
                        .setNegativeButton(R.string.no, null)
                        .show()
                }
            })
            //endregion

            //region [Handle imageViewTrash: delete member]
            val iVTrash = holder.imageViewTrash
            
            iVTrash.setColorFilter(member.color)
            iVTrash.setOnTouchListener(object : OnTouchImageViewListener(member.color) {
                override fun onTrigger() {
                    AlertDialog.Builder(activity)
                        .setMessage(String.format(resources.getString(R.string.delete_member_), member.name))
                        .setPositiveButton(resources.getString(R.string.ok)) { _, _ ->
                            if (!member.nodeId.isNullOrBlank())
                                member.remove(roomKey, OnCompleteListener<Void?> {
                                    Snackbar.make(requireActivity().findViewById(android.R.id.content), resources.getString(R.string.member__deleted, member.name), Snackbar.LENGTH_LONG)
                                        .setAction(resources.getString(R.string.undo)) { member.upload(roomKey) }
                                        .setActionTextColor(member.color / 3)
                                        .show()
                                })
                            else
                                Toast.makeText(context, resources.getString(R.string.an_error_has_occurred) + "member doesn't exist in database, try refreshing.",
                                    Toast.LENGTH_LONG).show()
                        }
                        .setNegativeButton(resources.getString(R.string.no), null)
                        .show()
                }
            })
            //endregion

            //region [Handle imageViewBrush: change member color]
            val iVBrush = holder.imageViewBrush
            iVBrush.setColorFilter(member.color)
            iVBrush.setOnTouchListener(object : OnTouchImageViewListener(member.color) {
                override fun onTrigger() {
                    if (group.memberList.size < colorList.size)
                        ColorPicker(activity).apply {
                            setTitle(resources.getString(R.string.choose_a_color_for_, member.name))
                            setColors(ArrayList( removeColorsFromArrayToHex (colorList) ))
                            
                            setOnFastChooseColorListener(object : OnFastChooseColorListener {
                                override fun setOnFastChooseColorListener(i: Int, color: Int) {
                                    if (color != Color.WHITE) {
                                        member.color = color
                                        member.upload(roomKey, null)
                                    }
                                }
        
                                override fun onCancel() {}
                            })
                            
                            setRoundColorButton(true)
                            show()
                        }
                    else
                        Snackbar.make(activity.findViewById(R.id.content), resources.getString(R.string.no_colors_are_available), Snackbar.LENGTH_LONG)
                            .show()
                }
            })
            //endregion
        }

        private fun removeColorsFromArrayToHex(list: IntArray) =
            list.map { if (checkColorAvailable(it)) it.toHex() else null }.filterNotNull()
    
        override fun getItemCount(): Int = if (this@MembersFragment::group.isInitialized) group.memberList.size else 0

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val imageViewMemberIcon: ImageView by itemView.bindView(R.id.imageViewMemberIcon)
            val imageViewMemberStar: ImageView by itemView.bindView(R.id.imageViewMemberStar)
            val textViewMemberName: TextView   by itemView.bindView(R.id.textViewMemberName)
            val imageViewLink: ImageView       by itemView.bindView(R.id.imageViewLink)
            val imageViewTrash: ImageView      by itemView.bindView(R.id.imageViewTrash)
            val imageViewBrush: ImageView      by itemView.bindView(R.id.imageViewBrush)
        }
    }
}